### ISSUE TITLE
nng_dial sync should not return until added to socket

### ISSUE THREAD

**Issue by gdamore** [nanomsg/nng#599]
The NNG dial operations are by default "synchronous", but unfortunately they are returning too soon.  In particular, we wind up completing these operations as soon as the lower layer transport is connected, but the pipe may not have actually been accepted by the socket.  This can lead to an application incorrectly assuming that it may begin sending traffic right away, when the socket doesn't yet have the pipe set up.

Basically we think we should revisit how this works to use aio's rather than cvs here, but some care and thought is needed.
